taglocker.go

```diff
@@ -0,0 +1,212 @@
+	// name? If so, just recurse on that.
+	if ok {
+		tn.mtx.RUnlock()
+		return child
+	}
+
+	// The child does not exist.  Upgrade to a writer lock in order
+	// to insert a new child into our set of children.
+	tn.mtx.RUnlock()
+	tn.mtx.Lock()
+
+	// Check again to see that nobody beat us to creating that child.
+	child, ok = tn.children[component]
+	if ok {
+		// Lucky us!
+		tn.mtx.Unlock()
+		return child
+	}
+
+	// Okay, we have no choice but to create the child ourselves.
+	child = newNode(component)
+	tn.children[component] = child
+
+	// Release our own lock and recurse on the child.
+	tn.mtx.Unlock()
+	return child
+}
+
+// lock traverses the slice of components, setting intention writer
+// locks along the branch of the lock tree, until it reaches the final
+// tag component.  There, it grabs an exclusive writer lock on that node.
+func (tn *tagLockerNode) lock(components []string) error {
+	// If we have no paths to traverse, lock ourselves!
+	if len(components) == 0 {
+		tn.tagLock.XLock()
+		return nil
+	}
+
+	// Otherwise, we are only part of the way to the final component
+	// to lock.  Take an intent read lock on this node.
+	tn.tagLock.IXLock()
+
+	currentComp := components[0]
+	remainingComp := components[1:]
+	return tn.getOrCreateChild(currentComp).lock(remainingComp)
+}
+
+// rLock traverses the slice of components, setting intention reader
+// locks along the branch of the lock tree, until it reaches the final
+// tag component.  There, it grabs a reader lock on that node.
+func (tn *tagLockerNode) rLock(components []string) error {
+	// If we have no paths to traverse, lock ourselves!
+	if len(components) == 0 {
+		tn.tagLock.SLock()
+		return nil
+	}
+
+	// Otherwise, we are only part of the way to the final component
+	// to lock.  Take an intent read lock on this node.
+	tn.tagLock.ISLock()
+
+	currentComp := components[0]
+	remainingComp := components[1:]
+	return tn.getOrCreateChild(currentComp).rLock(remainingComp)
+}
+
+// rUnlock unlocks a path that has already been previously
+// locked for reader access.  It decrements the reader count on
+// the final component in the path and removes a read intention
+// on all other paths.
+func (tn *tagLockerNode) rUnlock(components []string) error {
+	// If we have no paths to traverse, unlock ourselves!
+	if len(components) == 0 {
+		tn.tagLock.SUnlock()
+		return nil
+	}
+
+	currentComp := components[0]
+	remainingComp := components[1:]
+
+	// Unlock our children first - we want to unlock in the opposite
+	// order that we acquired the locks.
+	tn.mtx.RLock()
+	child, ok := tn.children[currentComp]
+	tn.mtx.RUnlock()
+	if !ok {
+		return fmt.Errorf("missing component %v", currentComp)
+	}
+	err := child.rUnlock(remainingComp)
+	if err != nil {
+		// TODO: what is the right thing to do here?  Should we
+		// attempt to unlock ourselves even if children failed to
+		// unlock correctly?  Either doing so or not doing so seems
+		// dangerous.
+		return err
```

Comment:
I'd probably prefix the error in this case: `errors.New("Failed to unlock children because %v", err.Error())`

Reply:
The right thing to do here might be for the caller to panic. We might want to have a public, typed error for this situation. If I were using this code and got an `UnrecoverableLockError`, I'd be inclined to panic.

Potentially returning a `MissingComponentError` on line 166 would help to distinguish these errors. I don't think a missing component is worth panicking about; that's something that can be handled.

Reply:
Makes sense to have typed errors.  I'll do so.

Reply:
Revisiting thos: Is there a way to force the caller to check the particular return type before panicing? I worry a caller will just propagate the error upwards and it won't every be dealt with beyond logging it, or whatever the top-level handler does?

I think I might push a patch that just panics() in the callee and see how that looks.

Reply:
You can't _force_ the caller to check the error type, but it's possible to provide a different error type that the caller can choose to check. (I recently pushed a branch with code that does that, including tests demonstrating use of `errors.Is` to check the type.)

That's the standard go way to do it. A `panic` would be considered bad practice. If a user doesn't check the returned error, they're writing bad code. I'd rather not follow bad practice in our code to try to prevent others from writing bad code.

We might want to create a `const ErrFatal = errors.New("fatal PLC error")` type, and then create `var ErrUnrecoverableLock = fmt.Errorf("Locking failed resulting in %w", ErrFatal)`. That way a user can do a generic check for `ErrFatal` if they're looking to see if they need to panic (with `if errors.Is(err, ErrFatal) {panic(err)}`) or a specific check for `ErrUnrecoverableLock` if they're looking for that one.

Reply:
My thinking was that since the only thing to possibly do in the case where the lock tree is corrupted is to panic, since allowing execution to continue would only lead either to a deadlock or to mutual exclusion being violated, it seems strictly more dangerous to defer that responsibility to the caller, especially since the "good practice" way of writing go is to just copy and paste the ol 'if err != nil { return err }` boilerplate without thinking about dispatching on a return type.  (Maybe Java had it right with checked exceptions after all)

If you feel strongly about it, I'll make the change (and document it in the docstring, which I should have done right off the bat), but I do worry about it being a potential footgun.

Reply:
Ok, agreed, a panic is a good choice. 

But I think you should panic on a public type in case the user wants to `recover` from it somehow. `panic(UnrecoverableError)` or something to that effect. I don't know if this is actually common practice, but it seems to me like a good practice? 🤷‍♀️

Reply:
I guess the problem with returning `UnrecoverableError` is that all of the panics would have the same string, and I see on your branch you're providing a different panic message in different places. Perhaps it could be a wrapped error type, so the caller could use `errors.Is(err, plc.UnrecoverableError)` but I'm not sure if there's a general thought on the appropriateness of passing an `err` to `panic`. It seems fine to me, but I haven't thought about this much.

Reply:
Good points.  Let me play around with that and see how it converges.